RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -11,40 +10,57 @@
+        return Flux.using(
+            () -> DirectoryReader.open(FSDirectory.open(indexDirectoryPath)),
+            reader -> {
+                logger.info(reader.maxDoc() + " documents found in the current Lucene index");
 
-        try (FSDirectory directory = FSDirectory.open(indexDirectoryPath);
-             IndexReader reader = DirectoryReader.open(directory)) {
-
-            // Add all documents to our output that have the _source field set and filled in
-            for (int i = 0; i < reader.maxDoc(); i++) {
-                Document document = reader.document(i);
-                BytesRef source_bytes = document.getBinaryValue("_source");
-                String id;
-                // TODO Improve handling of missing document id (https://opensearch.atlassian.net/browse/MIGRATIONS-1649)
+                return Flux.range(0, reader.maxDoc()) // Extract all the Documents in the IndexReader
+                           .map(i -> getDocument(reader, i))
+                           .cast(Document.class);
+            },
+            reader -> { // Close the IndexReader when done
                 try {
-                    id = Uid.decodeId(reader.document(i).getBinaryValue("_id").bytes);
-                } catch (Exception e) {
-                    logger.warn("Unable to parse Document id from Document");
-                    id = "unknown-id";
-                }
-                if (source_bytes == null || source_bytes.bytes.length == 0) { // Skip deleted documents
-                    logger.info("Document " + id + " is deleted or doesn't have the _source field enabled");
-                    continue;
+                    reader.close();
+                } catch (IOException e) {
+                    logger.error("Failed to close IndexReader", e);
                 }
+            }
+        ).filter(doc -> doc != null); // Skip docs that failed to read
+    }
 
-                documents.add(document);
-                logger.info("Document " + id + " read successfully");
+    private static Document getDocument(IndexReader reader, int docId) {
+        try {
+            Document document = reader.document(docId);
+            BytesRef source_bytes = document.getBinaryValue("_source");
+            String id;
+            try {
+                id = Uid.decodeId(reader.document(docId).getBinaryValue("_id").bytes);
+            } catch (Exception e) {
+                StringBuilder errorMessage = new StringBuilder();
+                errorMessage.append("Unable to parse Document id from Document.  The Document's Fields: ");
+                document.getFields().forEach(f -> errorMessage.append(f.name()).append(", "));
+                logger.error(errorMessage.toString());
+                return null; // Skip documents with missing id
+            }
+            if (source_bytes == null || source_bytes.bytes.length == 0) {
+                logger.warn("Document " + id + " is deleted or doesn't have the _source field enabled");
```

Comment:
Would this be better suited for `info` if this expected for deleted documents

Reply:
I felt (and I guess still feel) that `warn` is probably the right level.  It's something that we should highlight the occurence of without being an `error`, per se.

Reply:
Which one is it - can you tell the difference?  Could the reader of the log tell the difference?  Is there something in the beginning of the log that would give the user a clue?

If _source wasn't enabled, could this flood the logs?

Reply:
Is there any chance that docId could have PII in it?  The docId could be customer generated, right?  Or are they only internal ids that are separately mapped to the customer-given ones?

If they're customer driven, I'd push this to debug to promote the policy that no PII could be shown for INFO and above logs.  This feels like it isn't a great spot to be in.  I'm hoping that there's a way to show an identifier without risking divulging a customer value.